From 8ecd3429d3ca74eb1f30414bf2ea93d29a8aca2e Mon Sep 17 00:00:00 2001
From: Steve Dickson <SteveD@redhat.com>
Date: Thu, 18 Sep 2014 14:20:40 -0400
Subject: [fs] sunrpc: remove all refcounting of groupinfo from rpcauth_lookupcred

Message-id: <1411050057-12417-78-git-send-email-steved@redhat.com>
Patchwork-id: 96251
O-Subject: [RHEL7.1 PATCH 77/94] SUNRPC: remove all refcounting of groupinfo from rpcauth_lookupcred
Bugzilla: 1111708
RH-Acked-by: J. Bruce Fields <bfields@redhat.com>
RH-Acked-by: Benjamin Coddington <bcodding@redhat.com>

From: NeilBrown <neilb@suse.de>

current_cred() can only be changed by 'current', and
cred->group_info is never changed.  If a new group_info is
needed, a new 'cred' is created.

Consequently it is always safe to access
   current_cred()->group_info

without taking any further references.
So drop the refcounting and the incorrect rcu_dereference().

Signed-off-by: NeilBrown <neilb@suse.de>
Signed-off-by: Trond Myklebust <trond.myklebust@primarydata.com>
Upstream Id: 122a8cda6a9a3594c4737570f5cd81636bcd20cf
Bugzilla: 1111708
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/sunrpc/auth.c b/net/sunrpc/auth.c
index 1d84887..7eb6f00 100644
--- a/net/sunrpc/auth.c
+++ b/net/sunrpc/auth.c
@@ -592,14 +592,8 @@ rpcauth_lookupcred(struct rpc_auth *auth, int flags)
  memset(&acred, 0, sizeof(acred));
  acred.uid = cred->fsuid;
  acred.gid = cred->fsgid;
- if (flags & RPCAUTH_LOOKUP_RCU)
-  acred.group_info = rcu_dereference(cred->group_info);
- else
-  acred.group_info = get_group_info(((struct cred *)cred)->group_info);
-
+ acred.group_info = cred->group_info;
  ret = auth->au_ops->lookup_cred(auth, &acred, flags);
- if (!(flags & RPCAUTH_LOOKUP_RCU))
-  put_group_info(acred.group_info);
  return ret;
 }
 
-- 
1.7.1